keps/sig-node/20190530-commit-class.md

```diff
@@ -0,0 +1,405 @@
+under-commit the physical resources on a group of nodes by modifying the
+amount of resource advertised by kubelet to the scheduler.
+
+## Motivation
+
+Cost is a key quality metric for business value delivered by a Kubernetes
+cluster. Cluster cost is typically proportional to the number of nodes required
+to run the Pods present on the cluster, which is a function of the resources
+requested by those Pods.
+
+One traditional strategy for minimizing the number of nodes required to run the
+desired workloads is over-commit of resources. The key hypothesis is that load
+spikes between distinct workloads are typically uncoordinated, and so the same
+resource may be allocated multiple times to different workloads. Higher
+over-commit levels achieve better cost outcomes, but risk compromising service
+quality if the uncoordinated-spike hypothesis does not hold.
+
+Kubernetes currently provides for resource over-commit by allowing Pods to
+declare themselves 'Burstable' or 'Best Effort' via container resource
+request:limit ratios. This is a powerful approach, but achieving high
+utilization at the cluster level requires coordination across many authors of
+Pod specs, which is challenging in a 'Namespace as a Service' multi-tenant
+Kubernetes environment, or surprising when implemented by fiat as a mutating
+admission webhook or required usage of Vertical Pod Autoscaler.
+
+The key belief behind this KEP is that Pod resource request:limit ratios are
+the business of the owner of that Pod, and should be primarily concerned with
+the cost of running that particular Pod. The cost of the cluster as a whole is
+outside their scope; as such, Kubernetes should provide a "big hammer" for
+cluster operators to use their wider perspective on cluster utilization to
+reduce cost via platform-level over-commit policies.
+
+### Goals
+
+* Allow platform-level over- or under- subscription of resources on groups of
+  nodes.
+
+### Non-Goals
+
+* Help cluster operators choose appropriate commit settings.
+* Automatically adjust commit settings to maximize utilization.
+* Replace VPA, HPA, or any other Pod-level right-sizing/auto-scaling API.
+  These APIs are complimentary: they help Pod owners optimize their Pod
+  footprint, while cluster-level over- or under-commit helps cluster operators
+  optimize the cluster footprint.
+* Replace Cluster Autoscaling. This is also complimentary: Cluster Autoscaling
+  changes the number of nodes based on resource allocation, while this KEP
+  changes the density of nodes.
+
+## Proposal
+
+A cluster-scoped `CommitClass` API will be added to the node API group:
+
+```yaml
+kind: CommitClass
+apiVersion: node.k8s.io/v1alpha1
+metadata:
+    name: high-cpu-density
+spec:
+  selector:
+  - matchExpressions:
+    - key: beta.kubernetes.io/instance-type
+    operator: In
+    values:
+    - compute-optimized
+  resources:
+  - name: cpu
+    percent: 1000 # 10x over-commit
+  - name: memory
+    percent: 120
```

Comment:
I had figured percentages <100 would be interpreted as 'under-commit', so a value of 50 would result in advertising 1/2 the resource value actually present. It isn't as clear to me when you would want to do that -- perhaps if you know your 'burstable' workloads are all going to burst around the same time of day, you'd want to leave some extra breathing room on that node group.

Very open to suggestions -- the prototype I banged together makes no special provision about >100 or <100; but there's probably a fair case to make that 'under-commit' is a dangerous foot-gun. Certainly it is a bit shocking when you set the percent to '0' and see your number of CPUs go to zero :)

Reply:
FWIW, I was merely pointing out that a user will certainly try to put a value <100 so might as well think through what it means and whether it would be permitted. :)

Personally, I think 100 should be the minimum and the operator should use the existing kube-system and host reserved amounts for handling things like reserved resources.